File: doctor/download_refs_007.mjs
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const mdFile = path.join(__dirname, '博士论文开题报告20260524-007.md');
const outDir = path.join(__dirname, 'PDF007');
const manifestPath = path.join(outDir, 'download_manifest.json');

const text = fs.readFileSync(mdFile, 'utf8');
const sec7 = text.split('# **七、文献参考**')[1] ?? text;
const urlRe = /\[(https?:\/\/[^\]]+)\]\(https?:\/\/www\.google\.com\/search/g;
const urls = [...sec7.matchAll(urlRe)].map((m) => m[1].replace(/\\#/g, '#'));

fs.mkdirSync(outDir, { recursive: true });

function slugFromUrl(url, index) {
  const u = new URL(url);
  let name = u.pathname.endsWith('.pdf')
    ? u.pathname.slice(1)
    : `${u.hostname.split('.')[0]}_${u.pathname.replace(/^\//, '').replace(/\//g, '_')}`;
  name = name.replace(/[^\w.\-]+/g, '_').slice(0, 80);
  return `${String(index).padStart(2, '0')}_${name}.pdf`;
}

function candidateUrls(url) {
  const list = [url];
  const m = url.match(/arxiv\.org\/abs\/([\d.]+(?:v\d+)?)/);
  if (m) list.unshift(`https://arxiv.org/pdf/${m[1]}.pdf`);
  else if (url.includes('arxiv.org/abs/')) {
    const id = url.split('/').pop();
    list.unshift(`https://arxiv.org/pdf/${id}.pdf`);
  }
  return list;
}

function isPdf(buf) {
  return buf.length >= 4 && buf[0] === 0x25 && buf[1] === 0x50 && buf[2] === 0x44 && buf[3] === 0x46;
}

const headers = {
  'User-Agent':
    'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
  Accept: 'application/pdf,*/*',
};

const results = [];
for (let i = 0; i < urls.length; i++) {
  const url = urls[i];
  const dest = path.join(outDir, slugFromUrl(url, i + 1));
  console.log(`[${i + 1}/${urls.length}] ${url}`);
  const info = { index: i + 1, url, ok: false };

  for (const tryUrl of candidateUrls(url)) {
    try {
      const resp = await fetch(tryUrl, { headers, redirect: 'follow', signal: AbortSignal.timeout(60000) });
      if (!resp.ok) {
        info.error = `HTTP ${resp.status}`;
        continue;
      }
      const buf = Buffer.from(await resp.arrayBuffer());
      const ct = (resp.headers.get('content-type') || '').toLowerCase();
      if (!isPdf(buf) && !ct.includes('pdf')) {
        info.error = `not PDF (${ct || buf.length + ' bytes'})`;
        continue;
      }
      fs.writeFileSync(dest, buf);
      info.ok = true;
      info.file = path.basename(dest);
      info.bytes = buf.length;
      info.from = tryUrl;
      console.log(`  -> OK (${buf.length} bytes)`);
      break;
    } catch (e) {
      info.error = e.message;
    }
  }
  if (!info.ok) console.log(`  -> FAIL: ${info.error}`);
  results.push(info);
  await new Promise((r) => setTimeout(r, 500));
}

const ok = results.filter((r) => r.ok).length;
const summary = { total: urls.length, success: ok, failed: urls.length - ok, items: results };
fs.writeFileSync(manifestPath, JSON.stringify(summary, null, 2), 'utf8');
console.log(`\nDone: ${ok}/${urls.length} -> ${outDir}`);
console.log(`Manifest: ${manifestPath}`);

File: doctor/download_refs_007.py
#!/usr/bin/env python3
"""Download PDFs from literature URLs in 博士论文开题报告20260524-007.md"""

import json
import re
import time
from pathlib import Path
from urllib.parse import urlparse

import requests

MD_FILE = Path(__file__).parent / "博士论文开题报告20260524-007.md"
OUT_DIR = Path(__file__).parent / "PDF007"
MANIFEST = OUT_DIR / "download_manifest.json"

HEADERS = {
    "User-Agent": (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
        "AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
    ),
    "Accept": "application/pdf,*/*",
}

# Match markdown link text that is a direct URL (not google search)
URL_PATTERN = re.compile(r"\[(https?://[^\]]+)\]\(https?://www\.google\.com/search")


def slug_from_url(url: str, index: int) -> str:
    parsed = urlparse(url)
    path = parsed.path.strip("/").replace("/", "_")
    if path.endswith(".pdf"):
        name = path
    else:
        name = f"{parsed.netloc.split('.')[0]}_{path}" if path else parsed.netloc
    name = re.sub(r"[^\w.\-]+", "_", name)[:80]
    return f"{index:02d}_{name}.pdf"


def arxiv_pdf_url(url: str) -> str | None:
    m = re.match(r"https?://arxiv\.org/abs/([\d.]+(?:v\d+)?)", url)
    if m:
        return f"https://arxiv.org/pdf/{m.group(1)}.pdf"
    return None


def candidate_urls(url: str) -> list[str]:
    urls = [url]
    if url.endswith(".pdf"):
        return urls
    apdf = arxiv_pdf_url(url)
    if apdf:
        urls.insert(0, apdf)
    if "arxiv.org/abs/" in url:
        aid = url.rsplit("/", 1)[-1]
        urls.insert(0, f"https://arxiv.org/pdf/{aid}.pdf")
    return urls


def is_pdf_content(resp: requests.Response) -> bool:
    ct = (resp.headers.get("Content-Type") or "").lower()
    if "pdf" in ct:
        return True
    return resp.content[:4] == b"%PDF"


def download_one(session: requests.Session, url: str, dest: Path) -> dict:
    last_err = None
    for try_url in candidate_urls(url):
        try:
            resp = session.get(try_url, headers=HEADERS, timeout=60, allow_redirects=True)
            if resp.status_code != 200:
                last_err = f"HTTP {resp.status_code}"
                continue
            if not is_pdf_content(resp):
                last_err = f"not PDF (content-type: {resp.headers.get('Content-Type')})"
                continue
            dest.write_bytes(resp.content)
            return {"ok": True, "file": str(dest.name), "bytes": len(resp.content), "from": try_url}
        except Exception as e:
            last_err = str(e)
    return {"ok": False, "error": last_err or "unknown"}


def main():
    text = MD_FILE.read_text(encoding="utf-8")
    # Only section 7 URLs
    sec7 = text.split("# **七、文献参考**", 1)[-1]
    urls = URL_PATTERN.findall(sec7)
    OUT_DIR.mkdir(parents=True, exist_ok=True)

    session = requests.Session()
    results = []

    for i, url in enumerate(urls, 1):
        url = url.replace("\\#", "#")
        dest = OUT_DIR / slug_from_url(url, i)
        print(f"[{i}/{len(urls)}] {url}")
        info = download_one(session, url, dest)
        info["index"] = i
        info["url"] = url
        results.append(info)
        status = "OK" if info["ok"] else f"FAIL: {info.get('error')}"
        print(f"  -> {status}")
        time.sleep(0.5)

    ok = sum(1 for r in results if r["ok"])
    summary = {"total": len(urls), "success": ok, "failed": len(urls) - ok, "items": results}
    MANIFEST.write_text(json.dumps(summary, ensure_ascii=False, indent=2), encoding="utf-8")
    print(f"\nDone: {ok}/{len(urls)} saved to {OUT_DIR}")
    print(f"Manifest: {MANIFEST}")


if __name__ == "__main__":
    main()
